src/Storages/MergeTree/MergeTreeIndexConditionBloomFilter.cpp

```diff
@@ -302,30 +324,82 @@ bool MergeTreeIndexConditionBloomFilter::traverseASTIn(
     return false;
 }
 
+
+static bool indexOfCanUseBloomFilter(const ASTPtr & parent)
+{
+    if (!parent)
+        return true;
+
+    /// `parent` is a function where `indexOf` is located.
+    /// Example: `indexOf(arr, x) = 1`, parent is a function named `equals`.
+    if (const auto * function = parent->as<ASTFunction>())
+    {
+        if (function->name == "and")
+        {
+            return true;
+        }
+        else if (function->name == "equals"
+            || function->name == "greater" || function->name == "greaterOrEquals"
+            || function->name == "less" || function->name == "lessOrEquals")
+        {
+            if (function->arguments->children.size() != 2)
+                return false;
+
+            /// We don't allow constant expressions like `indexOf(arr, x) = 1 + 0` but it's neglible.
+            if (const ASTLiteral * left = function->arguments->children[0]->as<ASTLiteral>())
+            {
+                if (function->name == "equals" && left->value.get<Int64>() != 0)
```

Comment:
There are several issues in this code:
1. UB when casting Float to Int64: `indexOf(arr, x) = nan`
2. Wrong results when UInt64 overflowed to Int64 in `get<Int64>` call.

Reply:
FieldVisitorAccurateComparison should be used.